### PR TITLE
Support pausing auction

### LINKED ISSUE #24: Replace require to assert when checking the balance of highest bidder in EndAuction()
https://github.com/QuarkChain/quarkchain-contracts/blob/51a12058502b0fdf675772b8d5ae4c0841861dd0/contracts/NonReservedNativeTokenManager.sol#L129

Since we stopping withdrawing the balance of the highest bidder, we internally will make sure that the equation is true and thus "require" should be replaced by "assert" (require will be only used to check the inputs are valid, while assert is used to check whether the internal data structure is correct).

### PR DESCRIPTION
**Pull request by hanyunx**
As mentioned in #15.
Also, fixed #24, #35.

### CODE REVIEW

File: contracts/NonReservedNativeTokenManager.sol [hunk 1]
```diff
@@ -67,6 +69,20 @@ contract NonReservedNativeTokenManager {
         auctionParams.minPriceInQKC = _minPriceInQKC;
         auctionParams.minIncrementInPercent = _minIncrementInPercent;
         auctionParams.duration = _duration;
+        auctionParams.isPaused = false;
+    }
+
+    function pauseAuction() public {
+        require(msg.sender == supervisor, "Only account in whitelist can pause the auction.");
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: can use a modifier `onlySupervisor`

File: contracts/NonReservedNativeTokenManager.sol [hunk 2]
```diff
@@ -80,6 +96,7 @@ contract NonReservedNativeTokenManager {
     }
 
     function bidNewToken(uint128 tokenId, uint128 price, uint64 round) public payable {
+        require(auction.isPaused == false, "Auction is paused.");
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: `!auction.isPaused`

File: contracts/NonReservedNativeTokenManager.sol [hunk 3]
```diff
@@ -128,23 +145,16 @@ contract NonReservedNativeTokenManager {
     }
 
     function endAuction() public {
+        require(auction.isPaused == false, "Auction is paused.");
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: `!isPaused`

File: contracts/NonReservedNativeTokenManager.sol [hunk 4]
```diff
@@ -67,6 +69,20 @@ contract NonReservedNativeTokenManager {
...
+    }
+
+    function pauseAuction() public {
+        require(msg.sender == supervisor, "Only account in whitelist can pause the auction.");
+        auction.isPaused = true;
+    }
+
+    function resumeAuction() public {
+        require(msg.sender == supervisor, "Only account in whitelist can resume the auction.");
+        if (uint128(now) > endTime() && auction.startTime != 0) {
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: `canEnd`?

File: contracts/NonReservedNativeTokenManager.sol [hunk 5]
```diff
@@ -50,23 +51,40 @@ contract NonReservedNativeTokenManager {
     constructor (address _supervisor, bool _allowMint) public {
         supervisor = _supervisor;
         allowMint = _allowMint;
+        auction.isPaused = true;
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: add a comment here to note that the contract will not work at first

File: contracts/NonReservedNativeTokenManager.sol [hunk 6]
```diff
@@ -50,23 +51,40 @@ contract NonReservedNativeTokenManager {
...
     {
-        require(msg.sender == supervisor, "Only account in whitelist can set auction details.");
         require(
             auction.startTime == 0,
             "Auction setting cannot be modified when it is ongoing."
         );
         auctionParams.minPriceInQKC = _minPriceInQKC;
         auctionParams.minIncrementInPercent = _minIncrementInPercent;
         auctionParams.duration = _duration;
+        auction.isPaused = false;
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: no need to update `isPaused`, should require supervisor to call `resume`

File: contracts/NonReservedNativeTokenManager.sol [hunk 7]
```diff
@@ -13,7 +13,8 @@ contract NonReservedNativeTokenManager {
     }
 
     struct Auction {
-        uint64 round;
+        bool isPaused;
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: add a `isPaused` getter

File: contracts/NonReservedNativeTokenManager.sol [hunk 8]
```diff
@@ -50,23 +51,40 @@ contract NonReservedNativeTokenManager {
...
+        auction.isPaused = false;
+    }
+
+    function pauseAuction() public onlySupervisor {
+        auction.isPaused = true;
+    }
+
+    function resumeAuction() public onlySupervisor {
+        if (canEnd()) {
+            resetAuction();
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: add a comment saying the auction results are regarded as invalid

File: contracts/NonReservedNativeTokenManager.sol [hunk 9]
```diff
@@ -79,7 +102,12 @@ contract NonReservedNativeTokenManager {
         );
     }
 
+    function getIsPaused() public view returns (bool) {
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: `isPaused`

File: contracts/NonReservedNativeTokenManager.sol [hunk 10]
```diff
@@ -89,10 +117,10 @@ contract NonReservedNativeTokenManager {
 
         if (auction.startTime == 0) {
             // Auction hasn't started. Start now
-            auction.startTime = uint64(now);
+            auction.startTime = uint128(now);
         }
 
-        require(nativeTokens[tokenId].owner == address(0), "Token Id already exists");
+        require(!nativeTokens[tokenId].isTaken, "Token Id already exists");
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: separate this thing to a separate PR. we want PR to be as simple as possible

File: contracts/NonReservedNativeTokenManager.sol [hunk 11]
```diff
@@ -50,23 +52,44 @@ contract NonReservedNativeTokenManager {
     constructor (address _supervisor, bool _allowMint) public {
         supervisor = _supervisor;
         allowMint = _allowMint;
+
+        // The contract will not work at first unless setted up by the supervisor
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: *setted up*?

File: contracts/NonReservedNativeTokenManager.sol [hunk 12]
```diff
@@ -50,23 +52,44 @@ contract NonReservedNativeTokenManager {
...
-        require(msg.sender == supervisor, "Only account in whitelist can set auction details.");
         require(
             auction.startTime == 0,
             "Auction setting cannot be modified when it is ongoing."
         );
         auctionParams.minPriceInQKC = _minPriceInQKC;
         auctionParams.minIncrementInPercent = _minIncrementInPercent;
         auctionParams.duration = _duration;
+
+        resumeAuction();
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: why do we want to resume auction here? setting parameters should only do the job of **setting parameters**

Comment (hanyunx) [reply contracts/NonReservedNativeTokenManager.sol]: > no need to update `isPaused`, should require supervisor to call `resume`

Ah I misunderstood the comment above.  But I don't think it reasonable to require supervisor to call `resume` at the first start. It should _looks_ automatically to enable the auction once it is set up.

Comment (ninjaahhh) [reply contracts/NonReservedNativeTokenManager.sol]: I suggest the otherwise because of two reasons:

1. the functions should be loosely coupled - "Do One Thing and Do It Well" from unix philosophy - parameter setting should set the parameters, resuming auctions should resume the auction, instead of resuming the auction directly after setting the parameters. if we really want to have the effect, the setting parameter function should probably be called `setAuctionParametersAndStartAuction`
2. as for do we actually want to start the auction right after setting the parameters, I don't see any benefit combining them together as both of them need supervisor access (besides saving one function call). instead, my concern is the `resume` function may get more complicated in the future, thus coupling those two functions together may lead to more maintenance headache

File: contracts/NonReservedNativeTokenManager.sol [hunk 13]
```diff
@@ -13,27 +13,29 @@ contract NonReservedNativeTokenManager {
...
+        uint32 round;
         uint64 overtime;
         uint128 startTime;
         Bid highestBid;
     }
 
     struct AuctionParams {
         uint128 duration;
-        // Following are new token auction specific
+        // Following are new token auction specific.
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: remove this comment since it's not relevant anymore

File: contracts/NonReservedNativeTokenManager.sol [hunk 14]
```diff
@@ -13,27 +13,29 @@ contract NonReservedNativeTokenManager {
...
         uint64 minIncrementInPercent;
         uint64 minPriceInQKC;
     }
 
     struct NativeToken {
+        bool isTaken;
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: this should be in another PR?

File: contracts/NonReservedNativeTokenManager.sol [hunk 15]
```diff
@@ -69,6 +78,18 @@ contract NonReservedNativeTokenManager {
         auctionParams.duration = _duration;
     }
 
+    function pauseAuction() public onlySupervisor {
+        auction.isPaused = true;
+    }
+
+    function resumeAuction() public onlySupervisor {
+        if (canEnd()) {
+            // The auction result is regarded as an invalid.
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: *as invalid*

File: contracts/NonReservedNativeTokenManager.sol [hunk 16]
```diff
@@ -133,23 +159,17 @@ contract NonReservedNativeTokenManager {
...
         require(canEnd(), "Auction has not ended.");
-        require(
-            balance[auction.highestBid.bidder] >= auction.highestBid.newTokenPrice,
-            "Should have enough balance."
-        );
+        assert (balance[auction.highestBid.bidder] >= auction.highestBid.newTokenPrice);
+
         balance[auction.highestBid.bidder] -= auction.highestBid.newTokenPrice;
         nativeTokens[auction.highestBid.tokenId].owner = auction.highestBid.bidder;
+        nativeTokens[auction.highestBid.tokenId].isTaken = true;
```
Comment (ninjaahhh) [inline contracts/NonReservedNativeTokenManager.sol]: same here
